src/VisualStudio/CodeLens/ReferenceCodeLensProvider.cs

```diff
@@ -73,37 +73,78 @@ public ReferenceCodeLensProvider(Lazy<ICodeLensCallbackService> codeLensCallback
-            var hostGroupId = await _lazyCodeLensCallbackService.Value.InvokeAsync<string>(
-                this, nameof(ICodeLensContext.TryGetHostGroupIdAsync), arguments: null, cancellationToken).ConfigureAwait(false);
-            var serviceName = await _lazyCodeLensCallbackService.Value.InvokeAsync<string>(
-                this, nameof(ICodeLensContext.TryGetServiceNameAsync), arguments: null, cancellationToken).ConfigureAwait(false);
-            if (hostGroupId is null || serviceName is null)
+            while (true)
             {
-                return null;
+                await Task.Delay(TimeSpan.FromSeconds(1.5), _cancellationTokenSource.Token).ConfigureAwait(false);
+
+                ImmutableArray<Guid> keys;
+                lock (_dataPoints)
+                {
+                    keys = _dataPoints.Keys.ToImmutableArray();
+                }
+
+                var projectVersions = await _lazyCodeLensCallbackService.Value.InvokeAsync<ImmutableDictionary<Guid, string>>(
+                    this,
+                    nameof(ICodeLensContext.GetProjectVersionsAsync),
+                    new object[] { keys },
+                    _cancellationTokenSource.Token).ConfigureAwait(false);
+
+                lock (_dataPoints)
+                {
+                    foreach (var (projectGuid, newVersion) in projectVersions)
+                    {
+                        if (_dataPoints.TryGetValue(projectGuid, out var oldVersionedPoints)
+                            && newVersion != oldVersionedPoints.version)
+                        {
+                            foreach (var dataPoint in oldVersionedPoints.dataPoints)
+                                dataPoint.Invalidate();
+
+                            _dataPoints[projectGuid] = (newVersion, oldVersionedPoints.dataPoints);
+                        }
+                    }
+                }
             }
+        }
+
+        private void AddDataPoint(DataPoint dataPoint)
+        {
+            lock (_dataPoints)
+            {
+                var versionedPoints = _dataPoints.GetOrAdd(dataPoint.Descriptor.ProjectGuid, _ => (version: VersionStamp.Default.ToString(), dataPoints: new HashSet<DataPoint>()));
+                versionedPoints.dataPoints.Add(dataPoint);
 
-            var hostGroup = new HostGroup(hostGroupId);
-            var serviceDescriptor = new ServiceDescriptor(serviceName) { HostGroup = hostGroup };
+                if (_pollingTask is null)
+                {
+                    _pollingTask = Task.Run(PollForUpdatesAsync).ReportNonFatalErrorAsync();
+                }
+            }
+        }
 
-            return await _client.RequestServiceAsync(serviceDescriptor, cancellationToken).ConfigureAwait(false);
+        private void RemoveDataPoint(DataPoint dataPoint)
+        {
+            lock (_dataPoints)
+            {
+                if (_dataPoints.TryGetValue(dataPoint.Descriptor.ProjectGuid, out var points)
```

Comment:
When do we expect `_dataPoints.TryGetValue` or the `Remove(dataPoint)` to return false here?

Reply:
I would not expect that.